### ISSUE TITLE
Jasmine clock() and native Promise

### ISSUE THREAD

**Issue by piuccio** [jasmine/jasmine#710]
Any idea or suggestion on how to deal with native Promises when mocking the clock?

> You can make setTimeout or setInterval synchronous executing the registered functions only once the clock is ticked forward in time.

Take a look at this example

```js
describe('Promise', function () {
    var resolvedValue;

    beforeEach(function () {
        resolvedValue = null;
        jasmine.clock().install();
    });
    afterEach(function () {
        jasmine.clock().uninstall();
    });

    function getDataPromise () {
        new Promise(function (resolve, reject) {
            setTimeout(function () {
                resolve(12);
            }, 100);
        }).then(function (value) {
            resolvedValue = value;
        });
    }

    function getDataJQuery () {
        var deferred = new $.Deferred();
        setTimeout(function () {
            deferred.resolve(12);
        }, 100);
        deferred.promise().then(function (value) {
            resolvedValue = value;
        });
    }

    it('native promise', function () {
        // This is async, but I'd like to turn it into sync
        getDataPromise();

        jasmine.clock().tick(200);

        expect(resolvedValue).toBe(12);
    });

    it('jquery promise', function () {
        // This is async, but I'd like to turn it into sync
        getDataJQuery();

        jasmine.clock().tick(200);

        expect(resolvedValue).toBe(12);
    });
});
```

In the test I create a promise and using the `clock()` I try to make it synchronous. The test on jQuery passes, the one using native `Promise` fails because internally the browser makes sure that the `then` callback is always asynchronous.

This test is overly simplified, but assume I don't have access to the promise object (it's deep down in my modules), how can I test this?

One solution would be to have some API that allows me to access to the native `setTimeout`, say this

```js
    it('native promise', function (done) {
        getDataPromise();

        jasmine.clock().tick(200);

        // This access to the native setTimeout
        jasmine.clock().setTimeout(function () {
            expect(resolvedValue).toBe(12);

            done();
        }, 0);
    });
```

Or maybe
```js
    it('native promise', function (done) {
        getDataPromise();

        jasmine.clock().tick(200).then(function () {
            expect(resolvedValue).toBe(12);
            
            done();
        });
    });
```

**Comment by slackersoft** [comment]
Closing this as we haven't seen any response to this issue. It also looks a bit like a "how do I test?" question and not an issue with jasmine itself. For further questions of this nature, please use the jasmine-js@googlegroups.com mailing list.

**Comment by jonhartmann** [comment]
I'd like to add to that just saying this is a "how to test" problem is pretty lame - I've got Promises that resolve asynchronously based on time - its totally reasonable to think that Clock, the interface used to work with time-based async events,  would work in this scenario. I shouldn't have to jump through a lot of hoops to test something when your API seems to offer a simple solution.

**Comment by slackersoft** [comment]
The problem here is that Jasmine can't know how your promise library is going to go about completing its promises. Some always do it in a `setTimeout` others try to be clever and do other faster things first. This means there isn't something that Jasmine can have changed within itself to fix the issue but rather something you have solve within your own suite. To this end, it is more of "How do I test Promises in Jasmine?" question, and not a "Jasmine does this thing that should change". We try to keep Jasmine's issues list to the latter.

Hope this helps. Thanks for using Jasmine!

**Comment by jonhartmann** [comment]
@slackersoft so, setting aside that Promises are ES6 now and not necessarily coming from a library, it still doesn't make sense that Jasmine wouldn't work in simple cases for this. For example, I came to this post today because I have a scenario where I have a method returning a promise that represents a timeout pole for a variable that rejects after a certain number of checks... right now I have to do setTimeout in my tests to test this feature, the fact that I can't trade done() and setTimeout() for clock which solves those same problems in other situations and executes faster is just frustrating. In this situation there are no libraries outside of Karma/Jasmine. If I'd written this code with timeouts and a callback the clock would have been working fine, right? But instead when I drop in clock the test just refuses to work.
